### PR TITLE
When removing a competition event, also delete any registrations for that event.

### LINKED ISSUE #377: Registrations for nonexistant events
If someone removes an event from a competition, there can be people who have already registered for that event. This results in pretty funky behavior. Copy pasting an email with Oliver about this:

```
I, [2016-02-03T09:30:54.263721 #31398]  INFO -- : Started PATCH "/competitions/HungarianNationals2016/registrations/all" for 213.222.165.69 at 2016-02-03 09:30:54 +0000
D, [2016-02-03T09:30:54.265401 #31398] DEBUG -- :   ^[[1m^[[36mUser Load (0.3ms)^[[0m  ^[[1mSELECT  `users`.* FROM `users` WHERE `users`.`id` = 378  ORDER BY `users`.`id` ASC LIMIT 1^[[0m
I, [2016-02-03T09:30:54.266600 #31398]  INFO -- : Processing by RegistrationsController#update_all as HTML
I, [2016-02-03T09:30:54.266653 #31398]  INFO -- :   Parameters: {"utf8"=>"✓", "authenticity_token"=>"[FILTERED]", "registration-90584"=>"1", "registrations_action"=>"accept-selected", "competition_id"=>"HungarianNationals2016"}
D, [2016-02-03T09:30:54.279122 #31398] DEBUG -- :   ^[[1m^[[35mCompetition Load (11.8ms)^[[0m  SELECT  `Competitions`.* FROM `Competitions` WHERE `Competitions`.`id` = 'HungarianNationals2016' LIMIT 1
D, [2016-02-03T09:30:54.279651 #31398] DEBUG -- :   ^[[1m^[[36mCACHE (0.0ms)^[[0m  ^[[1mSELECT  `Competitions`.* FROM `Competitions` WHERE `Competitions`.`id` = 'HungarianNationals2016' LIMIT 1^[[0m  [["id", "HungarianNationals2016"]]
D, [2016-02-03T09:30:54.281138 #31398] DEBUG -- :   ^[[1m^[[35mUser Exists (0.2ms)^[[0m  SELECT  1 AS one FROM `users` INNER JOIN `competition_organizers` ON `users`.`id` = `competition_organizers`.`organizer_id` WHERE `competition_organizers`.`competition_id` = 'HungarianNationals2016' AND `users`.`id` = 378 LIMIT 1
D, [2016-02-03T09:30:54.281522 #31398] DEBUG -- :   ^[[1m^[[36mCACHE (0.0ms)^[[0m  ^[[1mSELECT  `Competitions`.* FROM `Competitions` WHERE `Competitions`.`id` = 'HungarianNationals2016' LIMIT 1^[[0m  [["id", "HungarianNationals2016"]]
D, [2016-02-03T09:30:54.282581 #31398] DEBUG -- :   ^[[1m^[[35mRegistration Load (0.2ms)^[[0m  SELECT  `Preregs`.* FROM `Preregs` WHERE `Preregs`.`competitionId` = 'HungarianNationals2016' AND `Preregs`.`id` = 90584 LIMIT 1
D, [2016-02-03T09:30:54.283848 #31398] DEBUG -- :   ^[[1m^[[36m (0.8ms)^[[0m  ^[[1mBEGIN^[[0m
D, [2016-02-03T09:30:54.284745 #31398] DEBUG -- :   ^[[1m^[[35mUser Load (0.2ms)^[[0m  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 378 LIMIT 1
D, [2016-02-03T09:30:54.285752 #31398] DEBUG -- :   ^[[1m^[[36mCACHE (0.0ms)^[[0m  ^[[1mSELECT  `Competitions`.* FROM `Competitions` WHERE `Competitions`.`id` = 'HungarianNationals2016' LIMIT 1^[[0m  [["id", "HungarianNationals2016"]]
D, [2016-02-03T09:30:54.286743 #31398] DEBUG -- :   ^[[1m^[[35m (0.1ms)^[[0m  ROLLBACK
I, [2016-02-03T09:30:54.287137 #31398]  INFO -- : Completed 422 Unprocessable Entity in 20ms (ActiveRecord: 13.3ms)
F, [2016-02-03T09:30:54.289123 #31398] FATAL -- :
ActiveRecord::RecordInvalid (Validation failed: Events invalid event ids: 333fm):
  app/controllers/registrations_controller.rb:104:in `block in update_all'
  app/controllers/registrations_controller.rb:102:in `each'
  app/controllers/registrations_controller.rb:102:in `update_all'
  lib/middlewares/warden_user_logger.rb:10:in `call'
  lib/middlewares/fix_accept_header.rb:12:in `call'
```

It looks like you were trying to update registration 90584, which corresponds to https://www.worldcubeassociation.org/registrations/90584/edit, Niki herself? Is it possible that Niki registered for 333 fewest moves, and then you removed the event?

I'm not sure what the best solution to this is. Instead of a validation error, we could just remove all invalid (no longer offered) events when you try to update a registration that is registered for an event that is not being offered.. When events are removed from competitions, should we go and update all the corresponding registrations to remove their registration for the event? I almost feel that we should not allow removal of an event unless nobody is registered for that event...

### PR DESCRIPTION
**Pull request by jfly**
(This fixes #377)

This was possible thanks to @pingskills's work in #894 to get rid of the `eventSpecs` column.

- I had to tweak our schema a bit to get this working. Right now, we have a `registration_events` table which joins `Registrations` with their events. However, I believe this makes it hard/impossible to associate `competition_events` with `registration_events` and add the `dependent: :destroy` we need. So, I added a migration to change our `registration_events` table to a `registration_competition_events` table. This feels like the right direction to go to me, although it did force me to add some hackiness to our `_associated_events_picker.html.erb`.
- I've created #925 to add a warning when people remove an event that has people registered for it. Is anyone opposed to merge this PR up without implementing #925?

### CODE REVIEW

**Comment by jonatanklosko** [comment]
What would happen with registrations that have only one event if we'd remove it?

**Comment by larspetrus** [comment]
I get stuck on the name `RegistrationCompetitionEvent`. I can't parse that in any way that makes sense. 

What actually _is_ a RegistrationCompetitionEvent? Can we come up with a better name?

I feel bad for not putting in the time to learn what an RCE _is_, and offering a name suggestion, but this is not the day for such things.

**Comment by jfly** [comment]
Registration is the table of people who have registered for competitions, no (person, competition) pair will appear twice in there.

Competition is the table of all competitions, one row per competition.

Event is the table of all wca events, I'm still not in love with it being in the database.

CompetitionEvent is a table that lets us know what events are being held at a competition.

RegistrationCompetitionEvent is a table that tells us what competition events a registration is for.

**Comment by larspetrus** [comment]
Aha, it's a join table!

Since it doesn't have any properties or behavior aside from joining, this looks more suited for `has_and_belongs_to_many` to me. Not that this is an exact science. See the **Many-to-many** section here: http://api.rubyonrails.org/v4.2.7/classes/ActiveRecord/Associations/ClassMethods.html

By that theory, the reason I struggle with naming this thing is that it _isn't_ actually a thing.

One thing that always makes me (mildly) confused about events is that both the 4x4x4 competition at Oslo Open 2017 and the abstract concept of 4x4x4 competitions are called "event" in the code. I'd be helped by separate terms.

File: WcaOnRails/app/models/registration_competition_event.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+class RegistrationCompetitionEvent < ActiveRecord::Base
```
Comment (larspetrus) [inline WcaOnRails/app/models/registration_competition_event.rb]: This class is probably not needed.

File: WcaOnRails/app/views/shared/_associated_events_picker.html.erb [hunk 2]
```diff
@@ -12,10 +12,17 @@ Variables:
...
 <%# Note: form_builder.object is the object having associated events. %>
 <% associated_events = form_builder.object.send(events_association_name)
    all_events = allowed_events.map do |event|
-     associated_events.find_by_event_id(event.id) || associated_events.build(event_id: event.id)
+     if using_competition_events
+       competition_event = form_builder.object.competition.competition_events.find_by!(event: event)
+       associated_events.find_by_competition_event_id(event.id) || associated_events.build(competition_event: competition_event)
+     else
+       associated_events.find_by_event_id(event.id) || associated_events.build(event_id: event.id)
+     end
```
Comment (larspetrus) [inline WcaOnRails/app/views/shared/_associated_events_picker.html.erb]: From usage, it looks like `form_builder.object` can be either a `Registration`, `User` or `Competition`.

I'd look at moving this code out to those classes. So there'd be an `.valid_events(allowed_events)` on each class, and this whole section simplifies into

`all_events = form_builder.object.valid_events(allowed_events)`

_(The method name is probably wrong)_

**Comment by jfly** [comment]
> Since it doesn't have any properties or behavior aside from joining, this looks more suited for has_and_belongs_to_many to me. Not that this is an exact science

Ah! I wondered if I'd ever run into HABTM. Unfortunately, I'm not sure it's possible to use HABTM along with `accepts_nested_attributes_for`. Googling [rails habtm nested attributes for](https://www.google.com/search?q=rails+habtm+accepts+nested+attributes+for&oq=rails+habtm+accept) has a lot of FUD, but I'm not really sure how this should work. Do you have experience using HABTM with accepts_nested_attributes_for?

> One thing that always makes me (mildly) confused about events is that both the 4x4x4 competition at Oslo Open 2017 and the abstract concept of 4x4x4 competitions are called "event" in the code. I'd be helped by separate terms.

I agree. In this PR, I tried to be consistent about the use of the term "event" vs "competition_event".

**Comment by jfly** [comment]
> What would happen with registrations that have only one event if we'd remove it?

@jonatanklosko, I've tested this by hand, and I can verify that it causes a registration to become invalid:

```ruby
irb(main):010:0> r = Registration.all.select { |r| r.events.length == 1 }.first
...
=> #<Registration id: 82439, competitionId: "MyComp4812018", name: "Enola Abbott Sr.", personId: "2016IGYA01", countryId: "United Kingdom", gender: "m", birthYear: 1990, birthMonth: 6, birthDay: 4, email: "coralie@steuber.org", guests_old: nil, comments: "Sit sit provident officia molestiae natus. Ut sequ...", ip: "1.1.1.1", user_id: nil, created_at: "2016-09-13 02:17:41", updated_at: "2016-09-13 02:17:41", guests: 3, accepted_at: nil>
irb(main):024:0> r.events
=> #<ActiveRecord::Associations::CollectionProxy [#<Event id: "333bf", name: "Rubik's Cube: Blindfolded", rank: 50, format: "time", cellName: "3x3 blindfolded">]>
irb(main):023:0> r.competition.competition_events.find_by(event_id: "333bf").destroy
irb(main):030:0> r.events.reload
irb(main):029:0> r.events
=> #<ActiveRecord::Associations::CollectionProxy []>
irb(main):025:0> r.valid?
  RegistrationCompetitionEvent Load (3.1ms)  SELECT `registration_competition_events`.* FROM `registration_competition_events` WHERE `registration_competition_events`.`registration_id` = 82439
=> false
irb(main):026:0> r.errors.messages
=> {:registration_competition_events=>["must register for at least one event"]}
```

I'm not sure what else we could do here. This makes me prefer not allowing removal of events that have people registered for them.

**Comment by larspetrus** [comment]
On Mon, Oct 10, 2016 at 6:41 PM, Jeremy Fleischman <notifications@github.com
> wrote:

> Do you have experience using HABTM with accepts_nested_attributes_for?


I've never used *accepts_nested_attributes_for* at all. If StackOverflow
says it doesn't work, they're very likely right.

Still, there is no need for a model class for the join table.

> I agree. In this PR, I tried to be consistent about the use of the term
"event" vs "competition_event".

Nice. My preference would be *event_type* vs *event*, but that is probably
disruptive.

-- 
Procrastinators - the Leaders of Tomorrow

File: WcaOnRails/app/views/shared/_associated_events_picker.html.erb (hunk 2 above)
Comment (jfly) [inline WcaOnRails/app/views/shared/_associated_events_picker.html.erb]: Good idea! I went with `events_to_associated_events`.

**Comment by jfly** [comment]
> Still, there is no need for a model class for the join table.

Do you mean I should be able to delete the `registration_competition_event` model and things should still work? When I do that, I get test errors from FactoryGirl like this:

```
  53) Registration requires quests >= 0
     Failure/Error: let(:registration) { FactoryGirl.create :registration }
     NameError:
       uninitialized constant Registration::RegistrationCompetitionEvent
     # ./spec/models/registration_spec.rb:5:in `block (2 levels) in <top (required)>'
     # ./spec/models/registration_spec.rb:74:in `block (2 levels) in <top (required)>'
     # -e:1:in `<main>'
```

> Nice. My preference would be *event_type* vs *event*, but that is probably disruptive.

Oh, I see. `event_type` would be one of the 18 WCA official events, and `event` would mean "an event type being run at a competition". I think I still prefer `event`/`competition_event` for now.

**Comment by larspetrus** [comment]
OK, your real experience overrides my theoretical prediction :)

Yeah, to my language instincts, an event is something that happens, like the 2x2x2 competition at Oslo Open 2017. And if you adopt that naming, the already existing `event` has to be renamed. Just one mans personal preference. I'm not claiming it's objectively better.

**Comment by jfly** [comment]
I just rebased this PR, and travis appears to still be happy.

If there are no objections, I plan to merge this up on Monday.

**Pull request merged**